packages/flutter_tools/lib/src/android/android_sdk.dart

```diff
@@ -409,6 +411,54 @@ class AndroidSdk {
     return null;
   }
 
+  // Returns the version of java in the format \d(.\d)+(.\d)+
+  // Returns null if version not found.
+  String? getJavaVersion({
+    required AndroidStudio? androidStudio,
+    required FileSystem fileSystem,
+    required OperatingSystemUtils operatingSystemUtils,
+    required Platform platform,
+    required ProcessUtils processUtils,
+  }) {
+    final String? javaBinary = findJavaBinary(
+      androidStudio: androidStudio,
+      fileSystem: fileSystem,
+      operatingSystemUtils: operatingSystemUtils,
+      platform: platform,
+    );
+    if (javaBinary == null) {
+      return null;
+    }
+    final RunResult result = processUtils.runSync(
+      <String>[javaBinary, '--version'],
+      environment: sdkManagerEnv,
+    );
+    if (result.exitCode != 0) {
+      globals.printTrace(
+          'java --version failed: exitCode: ${result.exitCode} stdout: ${result.stdout} stderr: ${result.stderr}');
+      return null;
+    }
+    return parseJavaVersion(result.stdout);
+  }
+
+  // Java versions have 2 formats
+  @visibleForTesting
+  static String? parseJavaVersion(String? rawVersionOutput) {
+    if (rawVersionOutput == null) {
+      return null;
+    }
+    // The contents that matter come in the format '11.0.18' or '1.8.0_202'.
+    final RegExp jdkVersionRegex = RegExp(r'(\d+\.\d+(\.\d+(_?\d+)?))');
```

Comment:
Nit: the `_?` could just be `_`, and would be less confusing that way since it wouldn't make the matching of a final number of 2 or more digits have different possible ways of matching.